### PR TITLE
Support returning data in a `Paginator` from option `resolver` in `@paginator` directive

### PR DESCRIPTION
**Pull request by Stevemoretz**
- [X] Added or updated tests
- [x] Documented user facing changes
- [X] Updated CHANGELOG.md

There is no user facing changes, this doesn't impact backward compatibility but adds a feature which I honestly thought it was included before I tried, then I saw it wasn't.

We can decide on adding an example to doc for it or not, but I didn't think it was a good idea to do it myself.

**Changes**

No changes...
You can now return an instance of LengthAwarePaginator from the builder of your paginate directive which would bypass all the builders, (I use this myself to paginate a simple array). eg:

```php
return new LengthAwarePaginator([
            [
                "ID" => "test",
                "display_name" => "ok",
                "arr" => [
                    "s" => 's'
                ]
            ],
            [
                "ID" => "test2",
                "display_name" => "ok",
                "arr" => [
                    "s" => 's'
                ]
            ],
        ],2, 15);
```

**Breaking changes**

No breaking changes.

### CODE REVIEW

**Comment by spawnia** [comment]
I am not against an option to return a paginator directly, I see how the schema transformation of `@paginate` is still useful. However, using `builder` for this seems wrong to me, I would rather see this as a new option - probably just called `resolver`. Then, the directive definition and docs can clearly describe that this option is mutually exclusive with `builder` and `model`, and that other options such as `scopes` will have no impact when it is used.

**Comment by Stevemoretz** [comment]
> I am not against an option to return a paginator directly, I see how the schema transformation of `@paginate` is still useful. However, using `builder` for this seems wrong to me, I would rather see this as a new option - probably just called `resolver`. Then, the directive definition and docs can clearly describe that this option is mutually exclusive with `builder` and `model`, and that other options such as `scopes` will have no impact when it is used.

Hello, thanks for responding to this PR. I'm new to lighthouse that's why this took sometime to do, the idea of using resolver instead of builder was great, I also had an idea to set the resolver as a default if App\Query\QueryName was present but needed to ask you wether you think that's a good idea or not? And if yes, a little guidance would be nice I don't know how to get the QueryName in App\Query\QueryName, and I'm pretty sure you got a helper method written for it.

**Comment by Stevemoretz** [comment]
Was going to try to implement nested support for it but amazingly it already got supported with what I wrote yesterday!

Just sharing a simple example:

```graphql
type Query {
    sum: D @paginate(resolver: "App\\GraphQL\\Queries\\Sum")
}

type D {
    ID: String
    display_name: String
    arr: S @paginate(resolver: "App\\GraphQL\\Queries\\Sum")
}

type S{
    s: String
    nested: T @paginate(resolver: "App\\GraphQL\\Queries\\Sum")
}

type T{
    s: String
}
```

./app/GraphQL/Queries/Sum.php

```php
<?php

namespace App\GraphQL\Queries;

use GraphQL\Type\Definition\ResolveInfo;
use Illuminate\Pagination\LengthAwarePaginator;
use Nuwave\Lighthouse\Support\Contracts\GraphQLContext;

final class Sum
{
    /**
     * @param  null  $_
     * @param  array{}  $args
     */
    public function __invoke($_, array $args, GraphQLContext $context, ResolveInfo $resolveInfo)
    {
        if($_){
            if($_["arr"] && count($resolveInfo->path) === 4){
                return $_["arr"];
            }

            if( $_["nested"] && count($resolveInfo->path) === 7){
                return $_["nested"];
            }
        }
        return new LengthAwarePaginator([
            [
                "ID" => "test",
                "display_name" => "ok",
                "arr" => new LengthAwarePaginator([
                    ["s" => 's', "nested" => new LengthAwarePaginator([
                        ["s" => 's2']
                    ],1,10)]
                ],1,10)
            ],
            [
                "ID" => "test2",
                "display_name" => "ok",
                "arr" => new LengthAwarePaginator([
                    ["s" => 's',"nested" => new LengthAwarePaginator([
                        ["s" => 's2']
                    ],1,10)]
                ],1,10)
            ],
        ],15, 15);
    }
}
```

----

Query:

```graphql
{
  sum(first: 50) {
    data {
      ID
      arr {
        paginatorInfo{
          count
        }
        data{
          s,
          nested{
            paginatorInfo{
              count
              currentPage
            }
            data{
              s
            }
          }
        }
      }
    }
  }
}
```

result:

```json
{
  "data": {
    "sum": {
      "data": [
        {
          "ID": "test",
          "arr": {
            "paginatorInfo": {
              "count": 1
            },
            "data": [
              {
                "s": "s",
                "nested": {
                  "paginatorInfo": {
                    "count": 1,
                    "currentPage": 1
                  },
                  "data": [
                    {
                      "s": "s2"
                    }
                  ]
                }
              }
            ]
          }
        },
        {
          "ID": "test2",
          "arr": {
            "paginatorInfo": {
              "count": 1
            },
            "data": [
              {
                "s": "s",
                "nested": {
                  "paginatorInfo": {
                    "count": 1,
                    "currentPage": 1
                  },
                  "data": [
                    {
                      "s": "s2"
                    }
                  ]
                }
              }
            ]
          }
        }
      ],
      "ID": 1
    }
  }
}
```

File: docs/5/api-reference/directives.md
Comment (spawnia) [inline docs/5/api-reference/directives.md:2435]: Update the `master` docs instead, see https://github.com/nuwave/lighthouse/blob/master/docs/.github/README.md#documenting-new-features

File: src/Pagination/PaginateDirective.php
Comment (spawnia) [inline src/Pagination/PaginateDirective.php:49]:
```suggestion
  Reference a function that resolves the field by directly returning data in a `LengthAwarePaginator` instance.
  Mutually exclusive with `builder` and `model`. Not compatible with  `scopes` and builder arguments such as `@eq`.
```

File: src/Pagination/PaginateDirective.php
Comment (spawnia) [inline src/Pagination/PaginateDirective.php:103]: This is fine, but let's also validate that only one of `model`, `builder` and `resolver` is used, else throw a `DefinitionException` indicating that they are mutually exclusive.

File: src/Pagination/PaginateDirective.php
Comment (spawnia) [inline src/Pagination/PaginateDirective.php:45]: Let's also amend this and `model` with a hint that they are mutually exclusive with each other.

**Comment by Stevemoretz** [comment]
Thanks for accepting the PR phpStan should be fine now.

File: src/Pagination/PaginateDirective.php
Comment (spawnia) [inline src/Pagination/PaginateDirective.php:130]: Actually, the return value of this method only says `Illuminate\Contracts\Pagination\Paginator` - I think it would not work with that though, given `PaginatorField` and `ConnectionField` expect a `LengthAwarePaginator`. So I think we should probably change the return value?

File: src/Pagination/PaginateDirective.php
Comment (spawnia) [inline src/Pagination/PaginateDirective.php:127]:
```suggestion
                $paginator = $this->getResolverFromArgument('resolver')($root, $args, $context, $resolveInfo);
```

Comment (spawnia) [reply src/Pagination/PaginateDirective.php:127]: The variable names you used are misleading.

File: tests/Unit/Pagination/PaginateDirectiveTest.php
Comment (spawnia) [inline tests/Unit/Pagination/PaginateDirectiveTest.php:725]:
```suggestion
        type Query {
            users: [User] @paginate(resolver: \"{$builder}\")
        }

        type User {
            id: ID
        }
        ");
```

File: tests/Unit/Pagination/PaginateDirectiveTest.php
Comment (spawnia) [inline tests/Unit/Pagination/PaginateDirectiveTest.php:754]:
```suggestion
            users: [Query!]! @paginate(resolver: "NonexistingClass@notFound")
```

**Review by spawnia — CHANGES_REQUESTED** [review]

**Comment by spawnia** [comment]
We have to keep `resolver` explicit for now, defaulting to classes in `App\GraphQL\Queries` would be a breaking change.

File: docs/5/api-reference/directives.md
Comment (Stevemoretz) [reply docs/5/api-reference/directives.md:2435]: Working on this after changing the master, wanted to revert v5 and run `yarn release` which I ran after yarn and yarn start, but it gives:

```
error Command "release" not found.
```

I checked package.json and there was no scripts for release indeed:

```json
{
    "license": "MIT",
    "dependencies": {
        "@vuepress/plugin-back-to-top": "1.8.2",
        "@vuepress/plugin-medium-zoom": "1.8.2",
        "fs-extra": "9.1.0",
        "vuepress": "1.8.2",
        "vuepress-theme-default-prefers-color-scheme": "1.0.9"
    },
    "devDependencies": {
        "prettier": "2.2.1"
    },
    "scripts": {
        "start": "vuepress dev .",
        "build": "vuepress build .",
        "prettify": "prettier --write --tab-width=2 **/**/*.md **/*.js .vuepress/*.js"
    }
}
```

Comment (Stevemoretz) [reply docs/5/api-reference/directives.md:2435]: > Update the `master` docs instead, see https://github.com/nuwave/lighthouse/blob/master/docs/.github/README.md#documenting-new-features

Working on this after changing the master, wanted to revert v5 and run `yarn release` which I ran after yarn and yarn start, but it gives:

```
error Command "release" not found.
```

I checked package.json and there was no scripts for release indeed:

```json
{
    "license": "MIT",
    "dependencies": {
        "@vuepress/plugin-back-to-top": "1.8.2",
        "@vuepress/plugin-medium-zoom": "1.8.2",
        "fs-extra": "9.1.0",
        "vuepress": "1.8.2",
        "vuepress-theme-default-prefers-color-scheme": "1.0.9"
    },
    "devDependencies": {
        "prettier": "2.2.1"
    },
    "scripts": {
        "start": "vuepress dev .",
        "build": "vuepress build .",
        "prettify": "prettier --write --tab-width=2 **/**/*.md **/*.js .vuepress/*.js"
    }
}
```

Comment (spawnia) [reply docs/5/api-reference/directives.md:2435]: Sorry, it is `make release` now. I just fixed the docs.

Comment (Stevemoretz) [reply docs/5/api-reference/directives.md:2435]: > Sorry, it is `make release` now. I just fixed the docs.

Thank you but is there still any alternatives for not using docker and releasing it?
I'm on MacOS and don't have docker. (Docker is terrible on MacOS...)

Comment (spawnia) [reply docs/5/api-reference/directives.md:2435]: `make release` requires `make`, not Docker. If you do not have that, just look into the makefile and see what it does - you will find it to be quite trivial 😉

Comment (Stevemoretz) [reply docs/5/api-reference/directives.md:2435]: > `make release` requires `make`, not Docker. If you do not have that, just look into the makefile and see what it does - you will find it to be quite trivial 😉

Sorry got confused by :

```
If you use Docker you can start up the environment (including docs) by running:

make setup
make node
```

Thought I need these first to run the release, ran it without these and it worked thanks!

File: src/Pagination/PaginateDirective.php
Comment (Stevemoretz) [reply src/Pagination/PaginateDirective.php:130]: I went through this now seems like `Illuminate\Pagination\LengthAwarePaginator` implements `Illuminate\Contracts\Pagination\LengthAwarePaginator` which extends `Illuminate\Contracts\Pagination\Paginator` so `Illuminate\Pagination\LengthAwarePaginator` is actually of typeof `Illuminate\Contracts\Pagination\Paginator` so I guess it made more sense if I had written:

```php
                assert(
                    $query instanceof Paginator,
                    "The method referenced by the resolver argument of the @{$this->name()} directive on {$this->nodeName()} must return a Paginator."
                );
```

instead, that means the return type doesn't need to be changed tried with `Illuminate\Pagination\Paginator` instead of `Illuminate\Pagination\LengthAwarePaginator` which also worked.

File: src/Pagination/PaginateDirective.php
Comment (Stevemoretz) [reply src/Pagination/PaginateDirective.php]: Copy paste error sorry :) 
Changed to $resolverResolver

File: src/Pagination/PaginateDirective.php
Comment (Stevemoretz) [reply src/Pagination/PaginateDirective.php]: For now I committed it the way I described above if you see anything wrong with it please let me know.

File: src/Pagination/PaginateDirective.php
Comment (Stevemoretz) [reply src/Pagination/PaginateDirective.php]: Done but I'm not sure if you like the message and the fact it's always mentioning model and builder I could make it show it conditionally but didn't want to make the code dirty! 
Also added tests for it.

Comment (spawnia) [reply src/Pagination/PaginateDirective.php]: I have had this issue a couple of time, please review my draft of how we can solve it once and for all: https://github.com/nuwave/lighthouse/pull/2233

File: src/Pagination/PaginateDirective.php
Comment (spawnia) [reply src/Pagination/PaginateDirective.php]: We should define the type to be as wide as possible, but as narrow as needed. The downstream resolvers require `Illuminate\Contracts\Pagination\LengthAwarePaginator`, so that is what we should expect.

Comment (Stevemoretz) [reply src/Pagination/PaginateDirective.php]: > We should define the type to be as wide as possible, but as narrow as needed. The downstream resolvers require `Illuminate\Contracts\Pagination\LengthAwarePaginator`, so that is what we should expect.

Wasn't aware that, that was used internally so does this look ok?

```php
        $fieldValue->setResolver(function ($root, array $args, GraphQLContext $context, ResolveInfo $resolveInfo): LengthAwarePaginator {
            if ($this->directiveHasArgument('resolver')) {
                $paginator = $this->getResolverFromArgument('resolver')($root, $args, $context, $resolveInfo);

                assert(
                    $paginator instanceof LengthAwarePaginator,
                    "The method referenced by the resolver argument of the @{$this->name()} directive on {$this->nodeName()} must return a LengthAwarePaginator."
                );

                return $paginator;
            }
```

File: src/Pagination/PaginateDirective.php
Comment (Stevemoretz) [reply src/Pagination/PaginateDirective.php:109]: > They are all exclusive with each other, only one may be given. Let's validate this with #2233 once it is finished.

That pull request looks very interesting! Please let me know when it's finished.

> For docs, consider the hints for mutually exclusive arguments in `CanDirective`.

Just to be clear are you talking about this?

```
  """
  Check the policy against the model instances returned by the field resolver.
  Only use this if the field does not mutate data, it is run before checking.

  Mutually exclusive with `query` and `find`.
  """
  resolved: Boolean! = false
```

Which I should use as a base to alter the hint I wrote for `@pagniate`

```
  """
  Reference a function that resolves the field by directly returning data in a `LengthAwarePaginator` instance.
  Mutually exclusive with `builder` and `model`. Not compatible with `scopes` and builder arguments such as `@eq`.
  Consists of two parts: a class name and a method name, seperated by an `@` symbol.
  If you pass only a class name, the method name defaults to `__invoke`.
  """
  resolver: String
```

File: src/Pagination/PaginateDirective.php
Comment (Stevemoretz) [reply src/Pagination/PaginateDirective.php]: Turned out making the return type LengthAwarePaginator is not a great idea, hen using the same directive without resolver using eloquent I get this error:

```
"debugMessage": "Nuwave\\Lighthouse\\Pagination\\PaginateDirective::Nuwave\\Lighthouse\\Pagination\\{closure}(): Return value must be of type Illuminate\\Contracts\\Pagination\\LengthAwarePaginator, Illuminate\\Pagination\\Paginator returned",
```

File: src/Pagination/PaginateDirective.php
Comment (Stevemoretz) [reply src/Pagination/PaginateDirective.php:109]: Got the tests working locally and LengthAwarePaginator actually fails a few tests, AbstractPaginator passes all tests.

**Comment by Stevemoretz** [comment]
@spawnia I haven't pushed my latest commit that uses `AbstractPaginator` (was waiting for your approval on it), please let me know if you want me to push it since I saw the tests failed but sound like for a different reason (conflict) but anyways... if they didn't fail there they would have later

File: src/Pagination/PaginateDirective.php
Comment (Stevemoretz) [reply src/Pagination/PaginateDirective.php:104]: @spawnia Used your PR now, everything looks great, thanks! Please check the latest commit.

File: src/Execution/Arguments/ArgumentSet.php
Comment (spawnia) [inline src/Execution/Arguments/ArgumentSet.php:160]: Those changes seem unrelated, please clean up.

Comment (Stevemoretz) [reply src/Execution/Arguments/ArgumentSet.php]: Oops messed how that got there from another branch :)

**Comment by Stevemoretz** [comment]
Changes requested
1 review requesting changes by reviewers with write access. 

@spawnia I can't find that change request, could you mention me there?

**Pull request merged**
